src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1396,6 +1418,32 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
 
   // helper methods
 
+  private def withAttributeNamespaceCheck[T](userInfo: UserInfo, attributeNames: Iterable[AttributeName])(op: => Future[T]): Future[T] = {
+    val namespaces = attributeNames.map(_.namespace).toSet
+
+    // anyone can modify attributes in the default namespace
+    if (namespaces == Set(AttributeName.defaultNamespace)) op
+    else {
+      val invalidNamespaces = namespaces -- AttributeName.validNamespaces
+
+      // all users are prevented from using invalid namespaces
+      if (invalidNamespaces.nonEmpty)
+        Future.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(message = s"Invalid attribute namespace(s) ${invalidNamespaces}", statusCode = StatusCodes.Forbidden)))
+      else {
+        // only curators can modify attributes in the library namespace
+        if (namespaces.contains(AttributeName.libraryNamespace))
+          tryIsCurator(userInfo.userEmail) flatMap { isCurator =>
+            if (isCurator) op
+            else Future.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(message = s"Must be library curator", statusCode = StatusCodes.Forbidden)))
+          }
+        else op   // not all default, yet all valid and no library; not currently reachable
```

Comment:
is it ever possible, given all the use cases that call this, for attributeNames to be the empty set? If so, this else condition is reachable.